qualtran/bloqs/arithmetic/comparison.py

```diff
@@ -1154,3 +1156,423 @@ def _clineardepthgreaterthan_example() -> CLinearDepthGreaterThan:
+            a, b = bb.add_from(self._half_greater_than_bloq.adjoint(), a=a, b=b, c=c, target=target)  # type: ignore
+        else:
+            a, b = bb.add_from(self._half_greater_than_bloq.adjoint(), a=b, b=a, c=c, target=target)  # type: ignore
+
+        return {'a': a, 'b': b}
+
+    def build_composite_bloq(
+        self,
+        bb: 'BloqBuilder',
+        a: 'Soquet',
+        b: 'Soquet',
+        c: Optional['Soquet'] = None,
+        target: Optional['Soquet'] = None,
+    ) -> Dict[str, 'SoquetT']:
+        if self.uncompute:
+            assert c is not None
+            assert target is not None
+            return self._uncompute(bb, a, b, c, target)
+        else:
+            assert c is None
+            assert target is None
+            return self._compute(bb, a, b)
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> 'BloqCountDictT':
+        extra_ops = {}
+        if isinstance(self.dtype, QInt):
+            extra_ops = {
+                SignExtend(self.dtype, QInt(self.dtype.bitsize + 1)): 2,
+                SignExtend(self.dtype, QInt(self.dtype.bitsize + 1)).adjoint(): 2,
+            }
+        if self.comparison_dtype in ('>=', '<='):
+            extra_ops[XGate()] = 1
+        adder = self._half_greater_than_bloq
+        if self.uncompute:
+            adder = adder.adjoint()
+        adder_call_graph = adder.build_call_graph(ssa)
+        assert isinstance(adder_call_graph, dict)
+        counts = {**adder_call_graph}
+        for k, v in extra_ops.items():
+            if k not in counts:
+                counts[k] = 0
+            counts[k] += v
+        return counts
+
+
+@frozen
+class _LinearDepthHalfComparisonBase(_HalfComparisonBase):
+    @cached_property
+    def _half_greater_than_bloq(self) -> Bloq:
+        return _HalfLinearDepthGreaterThan(self.dtype, uncompute=False)
+
+
+@frozen
+class LinearDepthHalfGreaterThan(_LinearDepthHalfComparisonBase):
+    r"""Compare two integers while keeping necessary ancillas for zero cost uncomputation.
+
+    Implements $\ket{a}\ket{b}\ket{0}\ket{0} \rightarrow \ket{a}\ket{b}\ket{b-a}\ket{a>b}$ using $n$ And gates.
+
+    This comparator relies on the fact that c = (b' + a)' = b - a. If a > b, then b - a < 0. We
+    implement it by flipping all the bits in b, computing the first half of the addition circuit,
+    copying out the carry, and keeping $c$ for the uncomputation.
+
+    Args:
+        dtype: dtype of the two integers a and b.
+        uncompute: whether this bloq uncomputes or computes the comparison.
+
+    Registers:
+        a: first input register.
+        b: second input register.
+        c: ancilla register that will contain $b-a$ and will be used for uncomputation.
+        target: A single bit output register to store the result of a > b.
+
+    References:
+        [Halving the cost of quantum addition](https://arxiv.org/abs/1709.06648).
+    """
+    comparison_dtype: str = attrs.field(default='>', init=False)
```

Comment:
I'm not sure this is how we want to do this sort of inheritance. It's a little confusing. And presumably this attribute will show up in the repr and when you tab complete on available attributes.

Reply:
how is this different from the ZeroState bloq? both set the value of a property of a base class and hide it

Reply:
but looks good now